misc/copyright.py

```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Taichi Authors. All rights reserved.
+# Use of this software is governed by the LICENSE file.
+"""
+Open each source file and add a copyright notice if it is missing.
+Could be faster with multiprocessing or async/await but this is
+just an one-off script on a fairly-sized project so it's unnecessary.
+"""
+
+import argparse
+import os
+import re
+import time
+import subprocess
+import sys
+from datetime import datetime
+from enum import Enum
+from typing import List
+
+# Sync with implementation
+# Do not use color sequences or hyperlink sequences so as to be portable on Windows.
+DOC_STRING = """Copyright notice inserter.
+
+Arguments:
+
+paths       For each file, it will be visited; for each directory, files under
+            that tree will be visited.
+
+--exts      Comma-separated extension names to restrict the files that to be
+            picked up, e.g. "cpp", "cpp,py,sh". If not given, all files on the
+            traversal paths with recognized extension names will be picked up.
+
+--dry-run   Do not modify the files (for development).
+
+--docs      Print this long documentation and exits.
+
+Effects:
+
+The scripts checks a file and modifies it if necessary.
+    (1) If a copyright notice is absent, it inserts it at the start.
+        As an exception, if there is a sharp-bang (#!) line at the first line,
+        the new notice is inserted after it.
+    (2) If a copyright notice is present, but is in a specific wrong format, it
+        modifies the notice: https://github.com/taichi-dev/taichi/issues/1024
+
+Examples:
+
+{file} .
+{file} taichi/common/core.cpp
+{file} benchmarks cmake docs examples misc python taichi tests
+{file} --exts "cpp,py" benchmarks cmake docs examples misc python taichi tests""".format(
+    file=os.path.relpath(__file__))
+
+
+class CommentStyle(Enum):
+    C_STYLE = 1  # /* .. */
+    CPP_STYLE = 2  # // .. => seems to be used in *.cu only
+    PY_STYLE = 3  # #..
+
+
+class FileActionResult(Enum):
+    INTACT = 1
+    INSERTED_NOTICE = 2
+    MODIFIED_NOTICE = 3
+
+
+# Sync with DOC_STRING
+FILE_EXT_TO_COMMENT_STYLES = {
+    ".h": CommentStyle.C_STYLE,
+    ".c": CommentStyle.C_STYLE,
+    ".cc": CommentStyle.C_STYLE,
+    ".cpp": CommentStyle.C_STYLE,
+    # Though CUDA is on top of C, existing files seem
+    # to prefer "//" comments.
+    ".cu": CommentStyle.CPP_STYLE,
+    ".py": CommentStyle.PY_STYLE,
+    ".sh": CommentStyle.PY_STYLE,
+}
+
+
+def get_ctime_year(filepath: str) -> str:
+    """
+    Get file creation time and return the year as a stringified integer.
+    Because git-clone overwrites ctime, we need to retrieve it
+    using git-log.
+    """
+    # %as: author date as a YYYY-MM-DD string
+    command = "git --no-pager log --reverse --format=\"%as\" {}".format(
+        filepath)
```

Comment:
For some reason I'm getting
```
/*******************************************************************************
     Copyright (c) %as" The Taichi Authors. All rights reserved.
     Use of this software is governed by the LICENSE file.
*******************************************************************************/
```
on my end. It seems that `%as` is not recognized by my git. Do you have the same behavior on your machine?

Reply:
Huh, this is only available on my Git tailored for the Chromium project.